compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,305 @@
+
+import core.Symbols._
+import core.Contexts.Context
+import core.Types._
+import core.Annotations.Annotation
+import core.Flags
+import core.Names.Name
+import core.StdNames.tpnme
+import ast.tpd._
+
+import collection.mutable
+
+import dotty.tools.dotc.{semanticdb => s}
+
+class TypeOps:
+  import SymbolScopeOps._
+  import Scala3.given
+  private val paramRefSymtab = mutable.Map[(LambdaType, Name), Symbol]()
+  private val refinementSymtab = mutable.Map[(RefinedType, Name), Symbol]()
+  given typeOps: TypeOps = this
+  extension (tpe: Type)
+    def toSemanticSig(using LinkMode, Context, SemanticSymbolBuilder)(sym: Symbol): s.Signature =
+      def enterParamRef(tpe: Type): Unit =
+        tpe match {
+          case lam: LambdaType =>
+            paramRefSymtab((lam, sym.name)) = sym
+
+          // for class constructor
+          // class C[T] { ... }
+          case cls: ClassInfo if sym.info.isInstanceOf[LambdaType] =>
+            val lam = sym.info.asInstanceOf[LambdaType]
+            cls.cls.typeParams.foreach { param =>
+              paramRefSymtab((lam, param.name)) = param
+            }
+
+          // type X[T] = ...
+          case tb: TypeBounds =>
+            enterParamRef(tb.lo)
+            enterParamRef(tb.hi)
+
+          case _ => ()
+        }
+
+      def enterRefined(tpe: Type): Unit =
+        tpe match {
+          case refined: RefinedType =>
+            val key = (refined, sym.name)
+            refinementSymtab(key) = sym
+
+          case rec: RecType =>
+            enterRefined(rec.parent)
+
+          case cls: ClassInfo
+            if (cls.cls.name == tpnme.REFINE_CLASS) =>
+            sym.owner.owner.info match {
+              // def foo(x: Parent { refinement }) = ...
+              case refined: RefinedType =>
+                enterRefined(refined)
+
+              // type x = Person { refinement }
+              case tb: TypeBounds =>
+                // tb = TypeBounds(
+                //   lo = RefinedType(...)
+                //   hi = RefinedType(...)
+                // )
+                enterRefined(tb.lo)
+                enterRefined(tb.hi)
+
+              // def s(x: Int): { refinement } = ...
+              case expr: ExprType =>
+                enterRefined(expr.resType)
+              case m: MethodType =>
+                enterRefined(m.resType)
+              case _ => ()
+            }
+          case _ => ()
+        }
+      enterParamRef(sym.owner.info)
+      enterRefined(sym.owner.info)
+
+      def loop(tpe: Type): s.Signature = tpe match {
+        case mt: MethodType =>
+          val stparams = Some(s.Scope())
+          val paramss =
+            if (sym.rawParamss.nonEmpty) sym.rawParamss else sym.paramSymss
+          val termParamss = paramss.filter(ps => ps.forall(!_.isTypeParam))
+          val sparamss = termParamss.map(_.sscope)
+          s.MethodSignature(
+            stparams,
+            sparamss,
+            mt.resType.toSemanticType(sym)
```

Comment:
we should use `finalResultType` here, this way we get the actual result type, and not a nested method type in the case of curried functions (a method type only has a single parameter list)
```suggestion
            mt.finalResultType.toSemanticType(sym)
```

Reply:
Fixed by https://github.com/lampepfl/dotty/pull/12885/commits/139491bf42806215007d785d7995ae9e6615b402
and then rewritten by https://github.com/lampepfl/dotty/pull/12885/commits/a8a137c36977125b3d65e2c2289268961ce2db14